src/valid/handles.rs

```diff
@@ -0,0 +1,337 @@
+                            crate::ImageQuery::Size { level } => level,
+                            crate::ImageQuery::NumLevels
+                            | crate::ImageQuery::NumLayers
+                            | crate::ImageQuery::NumSamples => None,
+                        })?;
+                    }
+                    &crate::Expression::Unary {
+                        op: _,
+                        expr: operand,
+                    } => {
+                        this_handle.check_dep(operand)?;
+                    }
+                    &crate::Expression::Binary { op: _, left, right } => {
+                        this_handle.check_dep(left)?.check_dep(right)?;
+                    }
+                    &crate::Expression::Select {
+                        condition,
+                        accept,
+                        reject,
+                    } => {
+                        this_handle
+                            .check_dep(condition)?
+                            .check_dep(accept)?
+                            .check_dep(reject)?;
+                    }
+                    &crate::Expression::Derivative {
+                        axis: _,
+                        expr: argument,
+                    } => {
+                        this_handle.check_dep(argument)?;
+                    }
+                    &crate::Expression::Relational { fun: _, argument } => {
+                        this_handle.check_dep(argument)?;
+                    }
+                    &crate::Expression::Math {
+                        fun: _,
+                        arg,
+                        arg1,
+                        arg2,
+                        arg3,
+                    } => {
+                        this_handle
+                            .check_dep(arg)?
+                            .check_dep_opt(arg1)?
+                            .check_dep_opt(arg2)?
+                            .check_dep_opt(arg3)?;
+                    }
+                    &crate::Expression::As {
+                        expr: input,
+                        kind: _,
+                        convert: _,
+                    } => {
+                        this_handle.check_dep(input)?;
+                    }
+                    &crate::Expression::CallResult(function) => {
+                        function.check_valid_for(functions)?;
+                    }
+                    &crate::Expression::AtomicResult { .. } => (),
+                    &crate::Expression::ArrayLength(array) => {
+                        this_handle.check_dep(array)?;
+                    }
+                }
+            }
+            Ok(())
+        };
+
+        let validate_function = |function: &_| {
+            let &crate::Function {
+                name: _,
+                arguments: _,
+                result: _,
+                ref local_variables,
+                ref expressions,
+                named_expressions: _,
+                body: _,
```

Comment:
Since many kinds of statements have expression handles (`Statement::Emit`, `Statement::If`) and function handles (`Statement::Call`), we'll need a recursive function to walk the statement tree and check those handles.

Reply:
Yep, this was a pretty big omission on my part, sorry. 😓 Pushed 23637853 and 75654ac7. Whaddaya think?

Reply:
It's all good, that's what reviews are for. (Pizza is tasty.)